src/main/java/mods/railcraft/common/fluids/FluidTools.java

```diff
@@ -66,6 +68,26 @@ public static String toString(@Nullable FluidStack fluidStack) {
         return fluidStack.amount + "x" + fluidStack.getFluid().getName();
     }
 
+    /**
+     * Forge is too picking here. So no {@link InvTools#isEmpty(ItemStack)} here.
+     *
+     * @param stack The stack to check
+     * @return True if the liquid is successfully consumed
+     */
+    public static boolean isSuccessfullyConsumed(@Nullable ItemStack stack) {
+        return stack != null && stack.stackSize == 0;
+    }
+
+    /**
+     * Forge is too picking here. So no {@link InvTools#isEmpty(ItemStack)} here.
+     *
+     * @param stack The stack to check
+     * @return True if the liquid failed to drain/fill
+     */
+    public static boolean isFailed(@Nullable ItemStack stack) {
```

Comment:
This one fixes `isEmpty()`'s weakness that it cannot discern null stacks from used one-time container of stack size zero.